### ISSUE TITLE
java8 features break compatibility with android sdk <24

### ISSUE THREAD

**Issue by masc3d** [OpenFeign/feign#774]
`feign-10` is now incompatible with sdk 24 or lower (when using eg. `JAXRSContract` which uses java8 lambdas), equalling 50% of android devices out there.

**Comment by masc3d** [comment]
throwing `NoClassDefFoundError` on
```
removeValues(consumes.value(), (mediaType) -> emptyToNull(mediaType) == null, String.class);
```

maybe it's not the lambda though, as those should be supported on all devices by now.

**Comment by masc3d** [comment]
caused by `java.util.function.Predicate` in `feign.Util` which is unsupported on sdk <24.

**Comment by masc3d** [comment]
> We made a conscious decision to move to Java 8 knowing that we would have some issues with Android

was there a public discussion I could read up on @kdavisk6?

well it probably implies that feign >= 10 will not be feasible to use for majority of android projects for what I would estimate to be 5-10 years until < sdk 24 devices become irrelevant.

a pretty harsh decision. I would rather have opted to move feign to kotlin to maintain portability and have better language features at the same time.

is it too late to re-evaluate this?

**Comment by kdavisk6** [comment]
It was discussed in #656.  We knew that this would be something that will come up.  The consideration towards Android was, let's see how it goes.  And now, here we are.    Moving to Kotlin is a more extreme move than moving to Java 8, and is something that we would encourage by creating a new library.  

Knowing that, what do you suggest we do?  From my perspective I see a few options.

- Restrict [Java 8 language usage to that of Android](https://developer.android.com/studio/write/java8-support)
- Require that Android developers use an SDK higher than 24 when using OpenFeign 10.x
- Create a `feign-android` specific version, that deals with this issue specifically.

We need to strike a balance between providing support for as many platforms as we can, while still moving the project forward, adding capabilities that allow us to do so.  Java, which has been stagnant for some time, is now moving very quickly forward.  From my perspective, which is just one of many, this is just the first step toward Java 10 and 11, which will bring another set of challenges.

Thoughts?

**Comment by masc3d** [comment]
> Restrict Java 8 language usage to that of Android

not feasible if you want to extensively leverage language features plus requires verification mechanisms to prevent unsupported java8 features from leaking in.

> Require that Android developers use an SDK higher than 24 when using OpenFeign 10.x

won't happen with > 50% unsupported atm and current rate of android moving forward. I cant imagine any project which can afford to ignore 50% or say even 25% of android devices.

> Create a feign-android specific version, that deals with this issue specifically.

lots of forking and maintenance. that could easily become more extreme than moving to kt.

**Comment by masc3d** [comment]
> Moving to Kotlin is a more extreme move than moving to Java 8, and is something that we would encourage by creating a new library.

I don't see it that extreme. can be applied incrementally as interoperability is excellent. I would rather go this route and get more balanced, elegant and powerful language instantly while preserving android compatibility.

then let's see where java is in 5 years. compared to kt its now (at least) 5 years behind.

**Comment by kdavisk6** [comment]
Kotlin is an different programming language.  While there is interoperability, it's akin to suggestion that we port to NodeJS, Python, or any other language.  It's the opinion of OpenFeign that large integrations/ports, such as the one you are proposing, be done as separate repositories.  See [HACKING](https://github.com/OpenFeign/feign/blob/master/HACKING.md), for more context on that.

With that being said, the only major change in 10.x is moving to Java 8.  Is there a feature in particular that is in 10.x that you would like to be able to use?

**Comment by masc3d** [comment]
> it's akin to suggestion that we port to NodeJS, Python

no it's not. those don't even relate to jvm or would compile to jvm bytecode without considerable footprint.

>  Is there a feature in particular that is in 10.x that you would like to be able to use?

no. not yet.

is android support relevant to you and will you accept PRs that remove java8 features incompatible with android for 10.x?

**Comment by kdavisk6** [comment]
Removing features that have already been accepted feels like bad form to me.  However, if you were to propose an alternative implementation of a feature, adding support for other platforms, that's something I think everyone could get behind.

And to make something clear, I speak for myself and myself alone.  There are other maintainers on this project with their own opinions and concerns.  They may agree or disagree with me on this issue.  Regardless, we made the decision to move to Java 8 on knowing the risks.  I understand and respect your position and your advocacy for continuing to support a wider range of Android platforms.  There is no reason, in my mind, that we cannot continue to do so, to a point.

It's not feasible for us (the maintainers), to impose platform restrictions on new features and contributions, as not all of those submitting changes are familiar or aware of those platform restrictions.  On the other hand, the community cannot expect us (the maintainers) to know about where this library is used all of the nuance that goes along with that.  We will continue to rely on the community to help here.

**Comment by masc3d** [comment]
> Removing features that have already been accepted feels like bad form to me.

@kdavisk6 I was referring to your initial suggestion `Restrict Java 8 language usage to that of Android`

> Are you interested in maintaining that branch?

@velo no, but I may kind of have to )

alright, so I understand your decision is set, that feign-10 will move forward without android support being (very) relevant. I will stick with feign-9 and we'll see how it goes.. 

Im open to discuss this further, but you can also close this issue, I will leave it to you.

**Comment by kdavisk6** [comment]
@masc3d I think you misunderstand.  No one here has stated that we do not want to support Android users.  What do need though is someone, like you, from the community to step up and help us do so.  That is what Open Source is all about.

**Comment by masc3d** [comment]
@kdavisk6 yes, but supporting android to a reasonable degree in feign-10 while moving to java8 implies serious trade-offs, like eg. not using `java.util.function`

that's why I asked if you would be accepting pull requests for maintaining android compatibility, like removing `java.util.function.Predicate` from `feign.core.Util`.

**Comment by kdavisk6** [comment]
That would be just fine with me.

**Comment by masc3d** [comment]
okay sounds good ) 

admittedly I cannot fully estimate consequences of eg compromising not to use `java.util.function` as Im 100% committed to kt for half a decade and another one to come. every time I have to fall back to java I literally squeal in agony. so more opinions are welcome (eg. @velo)

**Comment by velo** [comment]
I think the best plan to move forward is lay down here:
https://github.com/OpenFeign/feign/issues/774#issuecomment-414071856

We move `java.util.Function` to `feign.Function` and same for any other functional interfaces as needed. And retrolambda if required.

All we need are volunteers.

**Comment by masc3d** [comment]
> We move `java.util.Function` to `feign.Function` and same for any other functional interfaces as needed. And retrolambda if required.

I added `streamsupport` on my branch as it has all this. or would you still prefer to replicate into `feign.Function` @velo

**Comment by masc3d** [comment]
issue will be resolved with >=agp-4.0 which has [extended desugaring](https://androidstudio.googleblog.com/2019/10/android-studio-40-canary-1-available.html)

**Comment by masc3d** [comment]
> issue will be resolved with >=agp-4.0 which has [extended desugaring](https://androidstudio.googleblog.com/2019/10/android-studio-40-canary-1-available.html)

rather wishful thinking. now that agp-4 is finalized it becomes clear that new desugaring [does not cover all java-8 APIs](https://developer.android.com/studio/write/java8-support-table) used in feign eg some reflection features.

so the issue remains. feign-11 now requires even higher SDK level 26.

**Comment by velo** [comment]
> so the issue remains. feign-11 now requires even higher SDK level 26.

only feign-java11 module requires java 11. Everything else should be java 8.

**Comment by masc3d** [comment]
> > so the issue remains. feign-11 now requires even higher SDK level 26.
> 
> only feign-java11 module requires java 11. Everything else should be java 8.

yes, it's not related to `feign-java11`.

feign-11 for java-8 uses [Executable#getParameters() ](https://developer.android.com/reference/java/lang/reflect/Executable#getParameters()) which is not available on SDK level < 26 and unsupported for desugaring.
